Framework/Crystal/src/CalibrationHelpers.cpp

```diff
@@ -0,0 +1,165 @@
+
+using namespace Mantid::API;
+using namespace Mantid::Geometry;
+using namespace Mantid::Kernel;
+
+namespace Mantid {
+namespace Crystal {
+
+namespace {
+  constexpr double RAD_TO_DEG = 180. / M_PI;
+}
+
+/**
+ * Converts a Quaternion to a corresponding matrix produce Rotx*Roty*Rotz,
+ * corresponding to the order Mantid uses in calculating rotations
+ * @param Q The Quaternion. It will be normalized to represent a rotation
+ * @param Rotx The angle in degrees for rotating around the x-axis
+ * @param Roty The angle in degrees for rotating around the y-axis
+ * @param Rotz The angle in degrees for rotating around the z-axis
+ */
+void CalibrationHelpers::quatToRotxRotyRotz(const Quat Q, double &Rotx,
+                                            double &Roty, double &Rotz) {
+  Quat R(Q);
+  R.normalize();
+  V3D X(1, 0, 0);
+  V3D Y(0, 1, 0);
+  V3D Z(0, 0, 1);
+  R.rotate(X);
+  R.rotate(Y);
+  R.rotate(Z);
+  if (Z[1] != 0 || Z[2] != 0) {
+    double tx = atan2(-Z[1], Z[2]);
+    double tz = atan2(-Y[0], X[0]);
+    double cosy = Z[2] / cos(tx);
+    double ty = atan2(Z[0], cosy);
+    Rotx = (tx * RAD_TO_DEG);
+    Roty = (ty * RAD_TO_DEG);
+    Rotz = (tz * RAD_TO_DEG);
+  } else // roty = 90 0r 270 def
+  {
+    double k = 1;
+    if (Z[0] < 0)
+      k = -1;
+    double roty = k * 90;
+    double rotx = 0;
+    double rotz = atan2(X[2], Y[2]);
+
+    Rotx = (rotx * RAD_TO_DEG);
+    Roty = (roty * RAD_TO_DEG);
+    Rotz = (rotz * RAD_TO_DEG);
+  }
+}
+
+/**
+ * Updates the ParameterMap for NewInstrument to reflect the position of the
+ *source.
+ *
+ * @param newInstrument The instrument whose parameter map will be changed to
+ *reflect the new source position
+ * @param L0 The distance from source to sample (should be positive)
+ * @param newSampPos The relative shift for the new sample position
+ * @param pmapOld The Parameter map from the original instrument (not
+ *NewInstrument). "Clones" relevant information into the newInstrument's
+ *parameter map.
+ */
+void CalibrationHelpers::fixUpSampleAndSourcePositions(
+    boost::shared_ptr<const Instrument> newInstrument, double const L0,
+    const V3D newSampPos, DetectorInfo &detectorInfo) {
+  boost::shared_ptr<ParameterMap> pmap = newInstrument->getParameterMap();
+
+  IComponent_const_sptr source = newInstrument->getSource();
+  IComponent_const_sptr sample = newInstrument->getSample();
+
+  V3D samplePos = detectorInfo.samplePosition();
+  if (samplePos != newSampPos) {
+    detectorInfo.setPosition(*sample, newSampPos);
+  }
+  V3D sourceRelPos = source->getRelativePos();
+  V3D sourcePos = detectorInfo.sourcePosition();
+  V3D parentSourcePos = sourcePos - sourceRelPos;
+  V3D source2sampleDir = samplePos - source->getPos();
+
+  double scalee = L0 / source2sampleDir.norm();
+  V3D newsourcePos = newSampPos - source2sampleDir * scalee;
+  V3D newsourceRelPos = newsourcePos - parentSourcePos;
+
+  detectorInfo.setPosition(*source, newsourceRelPos);
```

Comment:
Need to set the absolute position here? If you do, can you get rid of the `instrument` argument?